### DISCUSSION TITLE
how to record with streamlink with codec x264/h264?

### DISCUSSION THREAD

**Discussion by ruzgatogosu** [streamlink/streamlink#4489] [Q&A] [answered]
is it even possible to use streamlink to record(download) streams with codec x264/h264?

**Comment by bastimeyer** [comment, marked as answer]
https://streamlink.github.io/latest/cli.html#file-output-options

Streamlink doesn't re-encode video/audio streams. Streams will at most get [re-muxed](https://en.wikipedia.org/wiki/Multiplexing#Video_processing), depending on the stream type or options passed. Most video streams however already are h264, so you don't have to re-encode anything. If you want to re-encode video (or audio) data, then you can pipe the output stream directly into FFmpeg and re-encode (and re-mux) on-the-fly, or you can do that afterwards after writing the stream data to the file system. This has been answered here a dozen times already.

**Reply by ruzgatogosu** [reply]
streamlink --hls-live-restart -o test1.mp4  --ffmpeg-fout mpegts --ffmpeg-video-transcode h264 --ffmpeg-audio-transcode aac https://www.twitch.tv/somestream best

i am sorry this doesnt work or am i doing something very wrong?

**Reply by bastimeyer** [reply]
`--ffmpeg-video-transcode` and `--ffmpeg-audio-transcode` are only relevant for streams which have to get remuxed by FFmpeg, eg. DASH streams or HLS streams with additional/external video/audio streams. These options do not transcode stream data of regular streams.

If you take a closer look at their descriptions, it says "When muxing streams". This isn't particularly great and should be improved.

As I said, how to pass data to FFmpeg has been answered here lots of times. You can do this via `--stdout`.

```sh
streamlink --stdout "$URL" "$QUALITY" | ffmpeg -i pipe:0 ...
```

But as I also already said, most streams are already h264. And in case of your example with Twitch, this is just nonsense, because the streams are h264+aac in an MPEG-TS transport-stream. If you want to **remux** to MPEG-4, see here:
https://github.com/streamlink/streamlink/discussions/4449#discussioncomment-2555845
